### ISSUE TITLE
DatasetView class breaks Liskov's rule.

### ISSUE THREAD

**Issue by owenlittlejohns** [pydata/xarray#8855]
### What is your issue?

Working on [migrating the datatree.py module](https://github.com/pydata/xarray/pull/8789) into `xarray/core` revealed that the `DatasetView` class, which implements `Dataset` while disabling methods to mutate the object, breaks [Liskov's substitution principle](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides). The type for one of the overloads of `DatasetView.__getitem__` is more general than the corresponding `Dataset.__getitem__` signature (due to the use of `Self` in [the Dataset signature](https://github.com/pydata/xarray/blob/main/xarray/core/dataset.py#L1529)).

```
# In Dataset:
class Dataset(...):
    ...
    @overload
    def __getitem__(self, key: Iterable[Hashable]) -> Self: ...
```

The use of `Self` means that signature inherited from the superclass has a return type of `DatasetView`, but the `DatasetView` signature is overridden to have a return type of `Dataset` (the more generalised parent).

To avoid this, a couple of implementations were attempted:

* A class that tries to intercept the methods that mutate the `Dataset` using `getattr`. This does not catch the `__setitem__` method, as it is a Magic Method, and those aren't affected by `getattr`.
* A `Metaclass` that can intercept Magic Methods, too. Implementation was inspired from [here](https://stackoverflow.com/a/9059858). I didn't get it to fully work, and eventually realised this was getting too complicated given the scope of the original problem.
* A mix-in for the mutating methods. I couldn't get this to work in the timescale agreed upon.
* Resorting back to ignoring the `mypy` errors for now, so we can proceed with the migration (given that there isn't a significant implementation concern identified from these type issues).

Also note, there is a tangentially-related known `mypy` error when a property setter accepts an argument of a different type to the property itself (https://github.com/python/mypy/issues/3004). This affects the assignment of `Dataset` objects to the `DataTree.ds` property. (Separate issue, but related)

**Comment by TomNicholas** [comment]
> A class that tries to intercept the methods that mutate the Dataset using `getattr`. This does not catch the `__setitem__` method, as it is a Magic Method, and those aren't affected by `getattr`.

This seems like it should be resolvable... But if we can satisfy mypy with some ignores we should definitely do that in the short term, we have much more interesting parts of datatree to get to, and we can always come back to this without making breaking changes.

**Comment by shoyer** [comment]
I think the right way to fix this from a typing perspective would be to make `Dataset` inherit from `DatasetView` instead of the other way around. Naming could be a little tricky, though!

**Comment by TomNicholas** [comment]
Would that work though? It would mean that `isinstance(dt.ds, Dataset)` would return `False`, which I wanted to avoid.

**Comment by shoyer** [comment]
> Would that work though? It would mean that `isinstance(dt.ds, Dataset)` would return `False`, which I wanted to avoid.
> 
> EDIT: But maybe that just isn't something we should be trying to make work.

Indeed, I think the situation is similar to Python's `dict` vs the general `collections.abc.Mapping` protocol. Notice how `MutableMapping` inherits from `Mapping`.

Another option for us could be to add an `immutable` or `frozen` attribute to `xarray.Dataset` itself.
